Implement a Debug Overlay for Enhanced Security

ronan-apple opened the pull request:
#### 06ca3f825ac470ac38ec38a3bf6fba4220fd9654
<pre>
Implement a Debug Overlay for Enhanced Security
<a href="https://bugs.webkit.org/show_bug.cgi?id=302001">https://bugs.webkit.org/show_bug.cgi?id=302001</a>
<a href="https://rdar.apple.com/164079715">rdar://164079715</a>

Reviewed by Per Arne Vollan.

Adds a Debug Overlay that indicates if a site has Enhanced Security enabled.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/EnhancedSecurityCocoa.h: Copied from Source/WebCore/page/DebugOverlayRegions.h.
* Source/WebCore/PAL/pal/cocoa/EnhancedSecurityCocoa.mm: Copied from Source/WebCore/page/DebugOverlayRegions.h.
(PAL::isEnhancedSecurityEnabledForCurrentProcessCached):
(PAL::isEnhancedSecurityEnabledForCurrentProcess):
(PAL::setEnhancedSecurityEnabledForCurrentProcess):
* Source/WebCore/page/DebugOverlayRegions.h:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::EnhancedSecurityOverlay::drawRect):
(WebCore::RegionOverlay::create):
(WebCore::DebugPageOverlays::updateOverlayRegionVisibility):
* Source/WebCore/page/DebugPageOverlays.h:
(WebCore::DebugPageOverlays::doAfterUpdateRendering):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/WebProcess.h:
* Tools/MiniBrowser/mac/SettingsController.h:
* Tools/MiniBrowser/mac/SettingsController.m:
(-[SettingsController _populateMenu:]):
(-[SettingsController enhancedSecurityOverlayVisible]):
(-[SettingsController preferenceKeyForRegionOverlayTag:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):

Canonical link: <a href="https://commits.webkit.org/302863@main">https://commits.webkit.org/302863@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e2e0b859fe9a8bd4a4d4a7f102364173dfd353f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130474 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2745 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137892 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132345 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2637 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99404 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e6c10a29-8887-422f-a084-1b263084af4d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133421 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2009 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116831 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80104 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0ef7e95e-5f17-4f64-abe5-d76de1de4cb0) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1932 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34961 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81151 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110493 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35465 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140371 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2535 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2316 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107919 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2579 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113172 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107828 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27445 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1968 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31615 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/55513 "The change is no longer eligible for processing.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2605 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65993 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2424 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2626 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2531 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebCore/page/DebugOverlayRegions.h:
```diff
@@ -33,6 +33,9 @@ enum class DebugOverlayRegions : uint8_t {
     EditableElementRegion = 1 << 3,
     InteractionRegion = 1 << 4,
+#if PLATFORM(COCOA)
+    EnhancedSecurity = 1 << 6,
```
Should this be `1 << 5`?

ronan-apple replied:
Ah, so these appear to be tied to equivalent values in WKPreferencesPrivate.h, which has:

```
WK_API_DEPRECATED_WITH_REPLACEMENT("ShowFrameProcessBordersEnabled", macos(15.0, WK_MAC_TBA), ios(18.0, WK_IOS_TBA), visionos(2.0, WK_XROS_TBA)) = 1 << 5,
```

In the place of `1 << 5`, but deprecated, so I think we should use `1 << 6` here still?

But perhaps this could use a comment for clarification of why the gap?

pvollan replied:
Good point! I think it would make sense with a comment.

ronan-apple replied:
Popped a comment in there now - thanks for checking!

pvollan commented on Source/WebCore/page/DebugPageOverlays.cpp:
```diff
@@ -674,6 +679,50 @@ bool InteractionRegionOverlay::mouseEvent(PageOverlay& overlay, const PlatformMo
+
+class EnhancedSecurityOverlay final : public RegionOverlay {
+public:
+    static Ref<EnhancedSecurityOverlay> create(Page& page)
```
Can this be `const Page&`?

ronan-apple replied:
I took a bit of an attempt here, but it doesn't look like this would be an easy addition, I think because of the way `RegionOverlay::create` and these subclasses are plumbed together. It may be feasible, but I think I'd have to plumb it all the way through all the overlay classes - do you think this should be done?

pvollan replied:
Thanks! I think it is OK to keep as-is.

pvollan commented on Source/WebCore/page/DebugPageOverlays.cpp:
```diff
@@ -674,6 +679,50 @@ bool InteractionRegionOverlay::mouseEvent(PageOverlay& overlay, const PlatformMo
+    {
+    }
+
+    bool updateRegion() override
```
Should we use `final`?

ronan-apple replied:
Yep, good point, updated!

pvollan commented on Source/WebCore/page/DebugPageOverlays.cpp:
```diff
@@ -674,6 +679,50 @@ bool InteractionRegionOverlay::mouseEvent(PageOverlay& overlay, const PlatformMo
+        m_overlay->setNeedsDisplay();
+        return true;
+    }
+    void drawRect(PageOverlay&, GraphicsContext&, const IntRect& dirtyRect) override;
```
Ditto.

ronan-apple replied:
Yep, good point, updated!

pvollan commented on Source/WebCore/page/DebugPageOverlays.h:
```diff
@@ -46,8 +46,16 @@ class DebugPageOverlays {
+    static constexpr unsigned NumberOfRegionTypes = static_cast<unsigned>(RegionType::EnhancedSecurity) + 1;
+#else
     static constexpr unsigned NumberOfRegionTypes = static_cast<unsigned>(RegionType::InteractionRegion) + 1;
+#endif
```
Can `EnhancedSecurity` be added to the enum for all platforms?

ronan-apple replied:
It could! We can't support it for all platforms as the `PAL` call to check if EnhancedSecurity is enabled only works on Cocoa, but we could add it to the enum and just have the `#ifdef` guarding trying to apply the region, if you think that'd be better?

pvollan replied:
Yes, I think that is preferable, IMHO. Thanks!

ronan-apple replied:
No worries - thanks for looking! I've removed the ifdef around these now, so it's a bit tidier. Hopefully that's okay?

ronan-apple replied:
Required a little more handling than this to build on other platforms (switch cases that needed to handle every case) - I've moved back to a previous approach we did where the only `#ifdef` part is the PAL portion that does the actual check on if we're in EnhancedSecurity. So on other platforms enabling this region will just always show as not having ES enabled.

pvollan commented on Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
```diff
@@ -39,6 +39,7 @@ typedef NS_OPTIONS(NSUInteger, _WKDebugOverlayRegions) {
     _WKEditableElementRegion = 1 << 3,
     _WKInteractionRegion WK_API_AVAILABLE(macos(13.0), ios(16.0)) = 1 << 4,
     _WKSiteIsolationRegion WK_API_DEPRECATED_WITH_REPLACEMENT("ShowFrameProcessBordersEnabled", macos(15.0, WK_MAC_TBA), ios(18.0, WK_IOS_TBA), visionos(2.0, WK_XROS_TBA)) = 1 << 5,
+    _WKEnhancedSecurity WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA), visionos(WK_XROS_TBA)) = 1 << 6,
```
Should this be `_WKEnhancedSecurityRegion`?

ronan-apple replied:
Ah yes, good spot, pushing this change now.

pvollan commented on Source/WebCore/page/DebugOverlayRegions.h:
```diff
@@ -33,6 +33,10 @@ enum class DebugOverlayRegions : uint8_t {
     EditableElementRegion = 1 << 3,
     InteractionRegion = 1 << 4,
+    // We must leave 1 << 5 empty due to prior use by SiteIsolation.
+#if PLATFORM(COCOA)
```
Can this #ifdef also be removed?

ronan-apple replied:
Yep, great point, thanks! That's the issue on the Linux builds, just pushing the fix.

Commit:
Implement a Debug Overlay for Enhanced Security
https://bugs.webkit.org/show_bug.cgi?id=302001
rdar://164079715

Reviewed by Per Arne Vollan.

Adds a Debug Overlay that indicates if a site has Enhanced Security enabled.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/cocoa/EnhancedSecurityCocoa.h: Copied from Source/WebCore/page/DebugOverlayRegions.h.
* Source/WebCore/PAL/pal/cocoa/EnhancedSecurityCocoa.mm: Copied from Source/WebCore/page/DebugOverlayRegions.h.
(PAL::isEnhancedSecurityEnabledForCurrentProcessCached):
(PAL::isEnhancedSecurityEnabledForCurrentProcess):
(PAL::setEnhancedSecurityEnabledForCurrentProcess):
* Source/WebCore/page/DebugOverlayRegions.h:
* Source/WebCore/page/DebugPageOverlays.cpp:
(WebCore::EnhancedSecurityOverlay::drawRect):
(WebCore::RegionOverlay::create):
(WebCore::DebugPageOverlays::updateOverlayRegionVisibility):
* Source/WebCore/page/DebugPageOverlays.h:
(WebCore::DebugPageOverlays::doAfterUpdateRendering):
* Source/WebKit/UIProcess/API/Cocoa/WKPreferencesPrivate.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/WebKit/WebProcess/WebProcess.h:
* Tools/MiniBrowser/mac/SettingsController.h:
* Tools/MiniBrowser/mac/SettingsController.m:
(-[SettingsController _populateMenu:]):
(-[SettingsController enhancedSecurityOverlayVisible]):
(-[SettingsController preferenceKeyForRegionOverlayTag:]):
* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController didChangeSettings]):

Canonical link: https://commits.webkit.org/302863@main
